Feature: add option to enable/disable "multimonitor focus follows mouse"

jtaala opened the pull request:
Resolves #389.

When using multiple monitors, by default, PaperWM focuses/activates a monitor when the mouse cursor enters said monitor.

This PR makes this behaviour a user settable option:
![image](https://github.com/paperwm/PaperWM/assets/30424662/76d79a5f-8f4b-409d-920b-284f1ca1e29c)

jtaala committed:
Added `monitor-focus-follows-mouse` option and implementation.

jtaala committed:
Added option to PaperWM settings.

jtaala committed:
FIXES: to properly focus/activation window (if exists).

jtaala committed:
Update selection when clicking to select new monitor.

jtaala committed:
FIXES: for drag n drop targets.

jtaala committed:
FX: for case where move drag window to another monitor but NOT on a drag
target.

jtaala committed:
When monitor focus follows mouse, inactivate selection when switching to
another monitor.

jtaala committed:
Replaced `add_actor` and `remove_actor` deprecated methods.

jtaala committed:
FIX: for case where switchWorkspace by clicking link (which opens window
in another space on another monitor) - if space is shown on that monitor
don't do gnome workspace switch animation.

Thesola10 commented:
This breaks touchscreen multi-monitor switching, regardless of what the setting is set to.

Thesola10 reviewed (dismissed):
Switching between monitors on touch is entirely broken (#655 behavior), regardless of what the "monitor focus follows mouse" setting is set to.

jtaala committed:
Added `touch-event` on space backvround to activate space.

Lythenas commented:
One thing I noticed it that the mouse move/position does not seem to be passed to windows on other monitors. I.e. I can hover over tabs on a firefox window that is not active in the current workspace but I can't hover over tabs on a firefox window on the other monitor.

This also means the first click is not passed to the application. So you always have to click twice, once to activate the other monitor and the second time to actually click where you want to click. I think this is a major inconvenience.

I'm not sure if this is something that is cause by this PR or not. But given that we no longer have a click overlay I would have expected this to work.

I also got the following error in the logs when dnd-ing windows (always happens from one monitor to another, sometimes happens when staying on one monitor). I checked terminator and files:

```
Jan 23 20:41:43 anarchy gnome-shell[45313]: meta_window_set_stack_position_no_sync: assertion 'window->stack_position >= 0' failed
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x5575082f4b40), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x5575082f4b40' has no handler with id '114818'
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x5575061c8930), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x5575061c8930' has no handler with id '114819'
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x557506208750), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x557506208750' has no handler with id '114820'
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x55750822ecb0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x55750822ecb0' has no handler with id '114821'
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x55750784c5a0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x55750784c5a0' has no handler with id '114822'
Jan 23 20:42:44 anarchy gnome-shell[45313]: Object Meta.BackgroundActor (0x55750b551ac0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
                                            == Stack trace for context 0x55750372eeb0 ==
                                            #0   7ffcc29ce5b0 b   resource:///org/gnome/gjs/modules/core/overrides/GObject.js:708 (14ed3819c1f0 @ 25)
                                            #1   7ffcc29ce660 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0ce0 @ 20)
                                            #2   7ffcc29ce760 b   self-hosted:203 (14ed38196a60 @ 245)
                                            #3   7ffcc29ce840 b   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:419 (2d116cfe0c90 @ 285)
                                            #4   5575037f9f98 i   file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/grab.js:354 (2d116cfea100 @ 73)
                                            #5   5575037f9f08 i   resource:///org/gnome/shell/ui/init.js:21 (14ed38170ba0 @ 48)
Jan 23 20:42:44 anarchy gnome-shell[45313]: ../glib/gobject/gsignal.c:2777: instance '0x55750b551ac0' has no handler with id '114823'
```

But I haven't noticed any issues, just the error in the log.

Also I noticed sometimes the window does not end up where I dropped it.

Not sure if all of this is caused by this PR, but I haven't seen that before. But then again I don't dnd windows.

jtaala committed:
Created create extra virtual click at cursor when
monitor-focus-follow-mouse is disabled (first click enables space, 2nd
virtual click happens after).

jtaala commented:
> This also means the first click is not passed to the application

I noticed this.  I'm not sure what to do about the position not being passed (it's not the ClickOverlay but I believe the background element eats events until the workspace is activated).

We can do something about needing two clicks through - I've added a virtual click afterwards (i.e. first click activates space, followed by another click on the actual cursor position.

Pull the latest to try it out.

jtaala commented:
P.S. I knew about the log errors - it's a signals cleanup issue after grab/drag finish, doesn't cause any actual issues so will look at that one later.

Thesola10 commented:
Oh, uh https://github.com/paperwm/PaperWM/pull/755/commits/21efa47d0443064ccd044bcde7190a1fc1f44fa9 did the trick it turns out, thanks @jtaala 

However, moving the mouse doesn't hand focus back to the display the mouse was on, as it did before. (I have enabled "focus follows mouse")

Lythenas commented:
While the first click is fixed.

> One thing I noticed it that the mouse move/position does not seem to be passed to windows on other monitors.

This also means that you e.g. can't drag firefox tabs between different firefox windows on different monitors. Instead it just creates a new window on the original monitor. This is a major inconvenience.

Also I use the Search Light extension. And I noticed it immediately closes again when I try to open it on an empty monitor. It works if I enable "Monitor focus follows mouse" or if a floating window is open anywhere or if the monitor has a tiled window.

jtaala committed:
FIX: monitor lastSpace check should be dynamic (ensures if space
activated by other means then pointer check remains valid).

tadfisher commented:
~~I think this is from #751, but hovering over any window element that causes the mouse cursor to change results in a constant 2Hz flickering between the changed cursor and the default one. For example, hovering over this comment box in Firefox.~~

Sorry, I'm on the release branch, so not applicable to this PR.

jtaala commented:
> Oh, uh [21efa47](https://github.com/paperwm/PaperWM/commit/21efa47d0443064ccd044bcde7190a1fc1f44fa9) did the trick it turns out, thanks @jtaala
> 
> However, moving the mouse doesn't hand focus back to the display the mouse was on, as it did before. (I have enabled "focus follows mouse")

Thanks @Thesola10 - I believe I know what that likely is.  Can you check the latest on this PR?

@Lythenas @Thesola10, it's looking like this approach to focusing (or not focusing monitors by mouse movement) might not be feasible as PaperWM tiling seems to need a space activated to properly focus elements (which is likely why the "monitor focus follows mouse" approach was implemented in the first place.

In any case, let's put this one on hold until we can get the touch stuff working again.

@Thesola10, let me know if the last commit fixed the issue with "moving mouse back to monitor not focusing" issue.  If so, we'll get those changes in first before looking at this PR again.

jtaala commented:
> ~I think this is from #751, but hovering over any window element that causes the mouse cursor to change results in a constant 2Hz flickering between the changed cursor and the default one. For example, hovering over this comment box in Firefox.~
> 
> Sorry, I'm on the release branch, so not applicable to this PR.

That's weird though - haven't seen that myself.  Create an issue for sure to see if others are seeing this (and def run `gather-system-info.sh`  to help debug that one).

Yeah, if you're on `release` then it wouldn't be #751 (I haven't merged that into `release` branch yet - it's still only in develop branch).

jtaala commented:
Hey all, see comment on why this PR's approach isn't currently feasible with PaperWM's multimonitor approach (not without some issues anyway):

https://github.com/paperwm/PaperWM/issues/389#issuecomment-1911089672

Thoughts?
